File: frontend/src/api/news.js
const API_BASE = ''

async function request(url, options = {}) {
  const res = await fetch(API_BASE + url, {
    ...options,
    headers: { 'Content-Type': 'application/json', ...options.headers },
  })
  const data = await res.json().catch(() => ({}))
  if (!res.ok) {
    const err = new Error(data.error || data.message || '请求失败')
    err.data = data
    throw err
  }
  return data
}

// ============ 新闻相关 API ============

export function getNewsList(params = {}) {
  const query = new URLSearchParams()
  if (params.days) query.set('days', String(params.days))
  if (params.category) query.set('category', params.category)
  if (params.source) query.set('source', params.source)
  if (params.limit) query.set('limit', String(params.limit))
  if (params.offset) query.set('offset', String(params.offset))
  return request('/api/news/latest?' + query.toString())
}

export function getNewsDetail(newsId) {
  const encoded = encodeURIComponent(newsId)
  return request('/api/news/detail/' + encoded)
}

export function syncNews() {
  return request('/api/news/sync', { method: 'POST' })
}

export function analyzeNews(params = {}) {
  return request('/api/news/analyze', {
    method: 'POST',
    body: JSON.stringify(params),
  })
}

export function getLatestAnalysis() {
  return request('/api/news/analysis/latest')
}

// ============ 市场数据 API ============

/**
 * 获取市场情绪数据
 * @param {number} days - 获取天数，默认7天
 */
export function getMarketSentiment(days = 7) {
  return request(`/api/market/sentiment?days=${days}`)
}

/**
 * 获取市场特征数据（涨跌停、成交额、换手率等）
 * @param {number} days - 获取天数
 */
export function getMarketFeatures(days = 30) {
  return request(`/api/market/features?days=${days}`)
}

/**
 * 获取宏观经济数据
 * @param {string} indicator - 指标类型
 * @param {number} days - 获取天数
 */
export function getMarketMacro(indicator = null, days = 30) {
  let url = '/api/market/macro?days=' + days
  if (indicator) {
    url += '&indicator=' + indicator
  }
  return request(url)
}

/**
 * 获取最新宏观经济数据
 * @param {string} indicator - 指标类型
 */
export function getMarketMacroLatest(indicator = null) {
  let url = '/api/market/macro/latest'
  if (indicator) {
    url += '?indicator=' + indicator
  }
  return request(url)
}

/**
 * 获取资金流向数据
 * @param {number} days - 获取天数
 */
export function getMarketMoneyFlow(days = 7) {
  return request(`/api/market/money-flow?days=${days}`)
}

/**
 * 获取全球市场数据
 * @param {number} days - 获取天数
 */
export function getMarketGlobal(days = 7) {
  return request(`/api/market/global?days=${days}`)
}

/**
 * 获取市场汇总数据
 */
export function getMarketSummary() {
  return request('/api/market/summary')
}

File: server/routes/news.py
# server/routes/news.py
"""
新闻相关 API 路由
"""

import logging

from flask import Blueprint, jsonify, request
from datetime import datetime

from data.news import NewsCrawler, NewsRepo
from analysis.llm import get_analyzer

logger = logging.getLogger(__name__)

news_bp = Blueprint("news", __name__, url_prefix="/api/news")


@news_bp.route("/detail/<path:news_id>", methods=["GET"])
def get_news_detail(news_id):
    """获取新闻详情"""
    import urllib.parse

    url = urllib.parse.unquote(news_id)

    repo = NewsRepo()
    news = repo.get_news_by_url(url)

    if not news:
        return jsonify({"code": 404, "message": "新闻不存在"})

    return jsonify(
        {
            "code": 0,
            "data": {
                "id": news.url,
                "title": news.title,
                "content": news.content,
                "source": news.source,
                "url": news.url,
                "published_at": news.published_at.isoformat()
                if news.published_at
                else None,
                "category": news.category,
            },
        }
    )


@news_bp.route("/list", methods=["GET"])
def get_news_list():
    """获取新闻列表"""
    days = request.args.get("days", 1, type=int)
    category = request.args.get("category", None)
    limit = request.args.get("limit", 100, type=int)

    repo = NewsRepo()
    news_list = repo.get_news(days=days, category=category, limit=limit)

    data = []
    for news in news_list:
        data.append(
            {
                "id": news.url,
                "title": news.title,
                "content": news.content[:200] if news.content else "",
                "source": news.source,
                "url": news.url,
                "published_at": news.published_at.isoformat()
                if news.published_at
                else None,
                "category": news.category,
            }
        )

    return jsonify({"code": 0, "data": data})


@news_bp.route("/latest", methods=["GET"])
def get_news_latest():
    """获取最新新闻"""
    repo = NewsRepo()
    news_list = repo.get_today_news()

    data = []
    for news in news_list[:10]:
        data.append(
            {
                "id": news.url,
                "title": news.title,
                "content": news.content[:200] if news.content else "",
                "source": news.source,
                "url": news.url,
                "published_at": news.published_at.isoformat()
                if news.published_at
                else None,
                "category": news.category,
            }
        )

    return jsonify({"code": 0, "data": data})


@news_bp.route("/sync", methods=["POST"])
def sync_news():
    """手动同步新闻"""
    crawler = NewsCrawler()
    repo = NewsRepo()

    if not crawler.can_fetch():
        return jsonify(
            {
                "code": 429,
                "message": "频率限制，请稍后再试",
                "status": crawler.get_fetch_status(),
            }
        )

    news_list = crawler.fetch_today()
    saved = repo.save_news(news_list)

    return jsonify(
        {
            "code": 0,
            "data": {
                "fetched": len(news_list),
                "saved": saved,
                "duplicates": len(news_list) - saved,
                "status": crawler.get_fetch_status(),
            },
        }
    )


@news_bp.route("/analyze", methods=["POST"])
def analyze_news():
    """分析新闻"""
    data = request.json or {}
    days = data.get("days", 1)
    use_deepseek = data.get("use_deepseek", True)

    repo = NewsRepo()
    news_list = repo.get_news(days=days, limit=50)

    news_dicts = []
    news_ids = []
    for news in news_list:
        news_dicts.append(
            {
                "title": news.title,
                "content": news.content,
                "source": news.source,
                "published_at": news.published_at.isoformat()
                if news.published_at
                else None,
                "category": news.category,
            }
        )
        if hasattr(news, "id"):
            news_ids.append(news.id)

    analyzer = get_analyzer()
    result = analyzer.analyze(news_dicts, use_deepseek=use_deepseek)

    repo.save_analysis(
        type(
            "AnalysisResult",
            (),
            {
                "news_count": result["news_count"],
                "summary": result["summary"],
                "deep_analysis": result["deep_analysis"],
                "market_impact": result["market_impact"],
                "key_events": result["key_events"],
                "investment_advice": result["investment_advice"],
                "analyzed_at": datetime.now(),
            },
        )()
    )

    if result.get("industry_tags") and news_ids:
        repo.update_news_industry_tags(news_ids, result["industry_tags"])

    return jsonify({"code": 0, "data": result})


@news_bp.route("/analysis/latest", methods=["GET"])
def get_analysis_latest():
    """获取最新分析结果"""
    repo = NewsRepo()
    result = repo.get_latest_analysis()

    if not result:
        return jsonify({"code": 0, "data": None, "message": "暂无分析结果"})

    return jsonify(
        {
            "code": 0,
            "data": {
                "news_count": result.news_count,
                "summary": result.summary,
                "deep_analysis": result.deep_analysis,
                "market_impact": result.market_impact,
                "key_events": result.key_events,
                "investment_advice": result.investment_advice,
                "analyzed_at": result.analyzed_at.isoformat()
                if hasattr(result.analyzed_at, "isoformat")
                else str(result.analyzed_at),
            },
        }
    )


@news_bp.route("/status", methods=["GET"])
def get_status():
    """获取爬虫状态"""
    crawler = NewsCrawler()
    repo = NewsRepo()

    return jsonify(
        {
            "code": 0,
            "data": {
                "crawler": crawler.get_fetch_status(),
                "news_count": repo.get_news_count(7),
                "categories": repo.get_categories(),
            },
        }
    )


@news_bp.route("/cleanup", methods=["POST"])
def cleanup_news():
    """清理过期新闻"""
    repo = NewsRepo()
    keep_days = request.json.get("keep_days", 30) if request.json else 30

    deleted = repo.cleanup_old_news(keep_days)

    return jsonify({"code": 0, "data": {"deleted": deleted}})
